src/conversion/model-to-view-converters.js

```diff
@@ -286,273 +314,163 @@ export function wrapItem( elementCreator ) {
- * {@link module:engine/conversion/model-to-view-converters~highlightText} method which uses default priority of
- * {@link module:engine/view/attributeelement~AttributeElement}.
- *
- * If the highlight descriptor will not provide `id` property, name of the marker will be used.
- * When `addHighlight` and `removeHighlight` custom properties are not present, element is not converted
- * in any special way. This means that converters will proceed to convert element's child nodes.
- *
- * @param {module:engine/conversion/model-to-view-converters~HighlightDescriptor|Function} highlightDescriptor
- * @return {Function}
- */
-export function highlightElement( highlightDescriptor ) {
-	return ( evt, data, consumable, conversionApi ) => {
-		const descriptor = typeof highlightDescriptor == 'function' ?
-			highlightDescriptor( data, consumable, conversionApi ) :
-			highlightDescriptor;
+		const viewRange = conversionApi.mapper.toViewRange( data.markerRange );
+		const type = evt.name.split( ':' )[ 0 ];
 
-		const modelItem = data.item;
+		// We will walk through every element in the range to highlight.
+		// Walking backwards will ensure that the walker won't break when change is applied inside range.
+		const treeWalker = new ViewTreeWalker( {
+			boundaries: viewRange,
+			startPosition: viewRange.end,
+			direction: 'backward'
+		} );
 
-		if ( !descriptor || data.markerRange.isCollapsed || !modelItem.is( 'element' ) ) {
-			return;
-		}
+		for ( const value of treeWalker ) {
+			// Ignore element start so that the same element is not returned twice.
+			if ( value.type == 'elementStart' ) {
+				continue;
+			}
 
-		if ( !consumable.test( data.item, evt.name ) ) {
-			return;
+			const viewItem = value.item;
+
+			// Will hold a new position for walker. It will depend on what action has been taken. See below.
+			let newPosition;
+
+			if ( viewItem.is( 'textProxy' ) ) {
+				// If the item is a text node, use text node highlighting helper.
+				// It will return a new position for walker.
+				newPosition = _highlightText( viewItem, highlightDescriptor, type, data, conversionApi );
+			} else if ( viewItem.is( 'containerElement' ) ) {
+				// If the item is container element, use element highlighting helper.
+				// Returned position depends on whether the element had custom highlighting enabled.
+				// If yes, returned position will be after the element.
+				// If no, no position will be returned, so the walker will continue inside the element.
+				newPosition = _highlightElement( viewItem, highlightDescriptor, type, data, conversionApi );
+			}
+			// Skip or step inside other elements (mostly attribute elements).
+
+			// If new position got specified, apply it to the walker.
+			if ( newPosition ) {
+				if ( newPosition.isBefore( viewRange.start ) ) {
+					// If the new position is beyond the range, stop processing.
+					break;
+				} else {
+					treeWalker.position = newPosition;
```

Comment:
`treeWalker.position` is at the moment readonly. We should check why.

Reply:
It seems that there is no real reason to have it read-only, other than it is read-only in `model.TreeWalker`. There, however, it makes more sense because we cache `this._visitedParent` -- `parent` of current `position` (it's for efficiency reasons).

However, I can propose an alternative solution for this function (not using `TreeWalker` at all).